tensorflow/lite/delegates/xnnpack/xnnpack_delegate.cc

```diff
@@ -3253,29 +3259,47 @@ TfLiteIntArray* Delegate::PrepareOpsToDelegate(TfLiteContext* context) {
             : static_cast<const char*>(input_tensor.data.data);
     switch (registration->builtin_code) {
       case kTfLiteBuiltinDequantize: {
-        if (input_tensor.type != kTfLiteFloat16) {
-          TF_LITE_KERNEL_LOG(
-              context, "unexpected tensor %d data type (%s) in node %d",
-              node->inputs->data[0], TfLiteTypeGetName(input_tensor.type),
-              producer_index);
-          TfLiteIntArrayFree(nodes_to_delegate);
-          return nullptr;  // Hard error.
-        }
-
         if (input_tensor.sparsity != nullptr) {
           TF_LITE_KERNEL_LOG(context,
-                             "unexpected FP16 sparse tensor %d in node %d",
+                             "unexpected FP16/INT8 sparse tensor %d in node %d",
                              node->inputs->data[0], producer_index);
           TfLiteIntArrayFree(nodes_to_delegate);
           return nullptr;  // Hard error.
         }
 
         // Actual data unpacking
-        float* unpacked_fp32_data = reinterpret_cast<float*>(unpacked_data);
-        const uint16_t* packed_fp16_data =
-            reinterpret_cast<const uint16_t*>(packed_data);
-        for (size_t i = 0; i < tensor_elements; i++) {
-          unpacked_fp32_data[i] = fp16_ieee_to_fp32_value(packed_fp16_data[i]);
+        switch (input_tensor.type) {
+          case kTfLiteFloat16:
+            DequantizeFloat16(reinterpret_cast<const uint16_t*>(packed_data),
+                              reinterpret_cast<float*>(unpacked_data),
+                              tensor_elements);
+            break;
+          case kTfLiteInt8: {
+            TfLiteAffineQuantization *quant_params =
+                static_cast<TfLiteAffineQuantization *>(input_tensor.quantization.params);
+
+            if (quant_params->scale->size != 1) {
+              TF_LITE_KERNEL_LOG(context,
+                                 "unsupported per channel dequantization for INT8 "
+                                 "tensor %d in XNNPACK delegate",
+                                 t);
+              return nullptr;  // Hard error.
```

Comment:
Missing "TfLiteIntArrayFree(nodes_to_delegate);" here?

Reply:
Right, thank you. Fixed